Implement a custom Emscripten File System which communicates with the JupyterLab Content Manager, giving file access to pyolite

martinRenou opened the pull request:
Fixes https://github.com/jupyterlite/jupyterlite/issues/665
Fixes https://github.com/jupyterlite/jupyterlite/issues/119

Progress towards https://github.com/jupyterlite/jupyterlite/issues/315

This PR implements a DriveFS custom file system which allows Emscripten kernels to have access to JupyterLab files. It uses a service worker to make any async task of accessing/deleting/renaming files into a synchronous task.

bollwyvl commented:
This seems like a fine starting point.

Note that, in case of emscripten-based kernels, we really should try to figure out a way to use the built-in `FS` capability, by mounting to a specific location... there are some links on #315. We gain a _lot_ from transparent read/write access to the same underlying system

bollwyvl commented:
> other drives

I haven't taken the time to look into other drives for some time... but yes, from the emscripten side, presumably we would need one `IDriveFS` implementation that could wrap any drive, each as another [mount point](https://emscripten.org/docs/api_reference/Filesystem-API.html#FS.mount) into the emscripten in-memory system. 

It's unfortunate drives are such a separate thing: as a litmus test, do they work in RetroLab?

> not limited to IndexedDB 

Strictly speaking, it's _not_ limited to indexeddb, as there are a lot of other localforage implementations that we don't have to implement.

The whole point of #315 is to be as consistent as possible in the UI and implementation. However many backends/frontends we need, they are all accessible through the same API, that work in the different apps (e.g. full lab, retro). So a new kernel only needs to implement against one contents API (if that, as #463 might normalize that, too, by necessity), while a new backend doesn't have to mess with service/web worker boundaries.

martinRenou commented:
> presumably we would need one IDriveFS implementation that could wrap any drive

In the case of https://github.com/jupyterlab-contrib/jupyterlab-filesystem-access the worker has no access to the files, unlike with indexeddb where any worker can access it. How would/Can an `IDriveFS` (in the emscripten world) work in that case?

Actually I see this JupyterLab drive API as being already the normalized FileSystem API for the JupyterLab world. JupyterLite does not do this at the moment, but it could technically use a custom drive backed by indexeddb.

It would be really nice if there was a way for the worker to mount such a custom JupyterLab drive. But it's unclear to me how this would work, as the worker lives in a different scope.

bollwyvl commented:
> How would/Can an IDriveFS (in the emscripten world) work in that case?

Right, it's tricky: and why it's not done/barely started yet.

On #463, I had started looking into using [comlink](https://github.com/GoogleChromeLabs/comlink), which does automatic proxies over worker boundaries.  Basically, instead of the API boundary being "this untyped `postMessage` pipe that kinda looks like jupyter kernel messaging" it would be, "this set of messages that can be awaited." Among them would of course be the kernel messages, but also "weird" stuff like interrupts (#460)... and contents messages that can efficiently ship large binary files.

> the JupyterLab world. 

Right: the question of #315 and #463 is: _what other robust worlds already exist_, e.g. emscripten, localforage, such that an implementer isn't starting over from scratch. And indeed, it doesn't seem like contents is an entirely solved problem: perhaps a yjs-on-wasm-FS would become viable, it might make sense to support, as it would be passing less content over the wire.

martinRenou commented:
Using comlink would indeed be very nice and make the codebase more readable

> and contents messages that can efficiently ship large binary files.

Do you think comlink is more efficient than `postMessage` for large binary files?

> perhaps a yjs-on-wasm-FS would become viable, it might make sense to support, as it would be passing less content over the wire.

Indeed. Though there is still this limitation with emscripten-fs that forces you to "sync" (with `syncfs`) the Python virtual filesystem if you want to open the files with the Python standard library, unless I missed something there?

Should I go ahead with trying to implement an approach that syncs the current drive with the Python virtual filesystem through `postMessage` calls (as proposed in the PR summary)? Although not ideal and not optimized, it could be a first version that works.

bollwyvl commented:
>  forces you to "sync" (with syncfs) the Python 

My understanding is that a mount point, once created, _is_ the system of record, instead of `MEMFS`, just like a unix mount point. So if you:
- uploaded a file to the file browser and `[*Path.glob("*")]` in the kernel you'd see the file listed
- wrote a file with `Path("foo.txt").write_text("foo")` it would show up, all with accurate modification times, etc.

> comlink is more efficient than postMessage for large binary files?

It's just sugar on postmessage, and one can override whatever they need, it's just moving away from _the huge `case` statement_... of course, it would _stay_ the huge `case` statement for kernel messages, but would probably be better typed. But crucially, we could add other things that _weren't_ kernel messages, and might have different call patterns.

As to performance: it would only actually transfer the file if the contents were requested.

>  go ahead with trying to implement an approach that syncs the current drive

I have been hesitant to do python-kernel-specific stuff at this low of a level: I really feel like solving it once at the webworker/emscripten level is really the right way to go. I tried a bit of `sync` stuff in the `emscripten-filesystem`, and didn't enjoy it very much.

martinRenou commented:
Thanks for your comments. I'm looking into comlink and emscripten-fs right now, see if I can come up with some kind of wrapper FS around the JupyterLab drives.

martinRenou commented:
I was able to use comlink for the kernel protocol 🎉 Now let's see if I can setup some drive/emscripten-fs interoperability

bollwyvl commented on packages/pyolite-kernel/src/kernel.ts:
```diff
@@ -192,23 +202,20 @@ export class PyoliteKernel extends BaseKernel implements IKernel {
     content: KernelMessage.ICompleteRequestMsg['content']
   ): Promise<KernelMessage.ICompleteReplyMsg['content']> {
-    return await this._sendRequestMessageToWorker('complete-request', content);
+    return this._workerKernel.complete(content, this.parent);
```
i personally like leaving the `await` in, even if it's just returning, as it will give more precise error stacks.

martinRenou replied:
Sure I'll add it back, thanks for the review.

I've been wondering if we could not even have the "workerKernel" be the one that implements the `IKernel` interface, because the interfaces become very similar now. This would need some restructure.

bollwyvl commented on packages/pyolite-kernel/src/worker.ts:
```diff
@@ -22,6 +22,10 @@ let stderr_stream: any;
 
+// TODO Import this locally
+declare var Comlink: any;
+importScripts("https://unpkg.com/comlink/dist/umd/comlink.js");
```
yeah: `importScripts` has a a couple nast complications, and an unpkg URL is definitely a non-starter.

This has some of the cleanest comlink worker stuff i was able to muster:

https://github.com/nrbgt/janki/tree/gh-6-add-cards/js/jupyterlab-sqlite3

there's some webpack conf needed, but then it can just use `file-loader` (maybe `assets` now?) and and plain old `import`s.

martinRenou replied:
Great thanks, I wasn't sure how to tackle this, I'll have a look at what you did. Webpack is a beast.

martinRenou commented:
One difficulty with the drive/emscripten-fs interoperability we might face is the fact that our custom emscripten-fs will be async for most (all) tasks. There isn't so much docs about custom emscripten FileSystem implementations and those I found were all synchronous.

martinRenou commented:
Thanks :) I was actually reading all those stuffs this morning when looking for docs implementing custom file systems.

> as for specific pyodide features

What do you see as being specific to pyodide? It looks to me that if we're able to have a emscripten-fs that can talk to a jupyterlab drive we could reuse the same implementation for all emscripten kernels?

bollwyvl commented on packages/pyolite-kernel/src/worker.ts:
```diff
@@ -26,6 +26,11 @@ let resolveInputReply: any;
 importScripts('https://unpkg.com/comlink/dist/umd/comlink.js');
 
+enum Mode {
+  file = 32768,
```
it's worth indicating this is a decimal mapping of an octal, e.g. https://github.com/gzuidhof/starboard-python/blob/88327bb56f7e2f673cf1ea4c5ff1e0a68c50e288/src/worker/emscripten-fs.ts#L67

martinRenou commented:
The code is a mess, but I have an implementation for `readdir` that works 😅 

![EMFS](https://user-images.githubusercontent.com/21197331/171868879-15305b40-1aa5-45d0-926f-e3d95c1dcc55.png)

bollwyvl commented:
Looks like this is going well... though the pain of not being able to use `import` properly is starting to catch up, as I'd think we don't want the drivefs to have to live forever inside the kernel.

Among the goals of #463 (which this PR is already starting, so here we go) would be: wrapper kernels (e.g. of `pyolite`: `pidgylite`, `robotlite`) would be only have to re-implement a small surface area to get new features like this (e.g. just the startup and kernel info) than the current required fork-and-patch.

- normalize the interface of the `IKernelWorker`
  - instead of the generic object
- import the concrete implementation in _another_ file that actually does the wrapping

So my initial play, loosely derived from that `sqlite` was something like this:

- `kernel.ts`
```ts
//  ...
import Comlink from 'comlink';
//  ...
export class PyoliteKernel extends BaseKernel implements IKernel {
  constructor(options: PyoliteKernel.IOptions) {
    super(options);
    this.initWorker(options); // no more string banging!
  }

  async initWorker(): Promise<void> {
    const this._rawWorker = new Worker(new URL('./worker-wrapper.js', import.meta.url), {
      type: 'module',
    });
    const this._remote = Comlink.wrap(this._worker);
    await this._remote.initialize(options);
    this._ready.resolve();
  }
  // ..
  private _worker: Worker;
  private _remote:  Comlink.Remote<IPyoliteKernelWorker>; 
}
```


- `worker-wrapper.ts`
```ts
// worker-comlink-wrapper.ts
import { expose } from 'comlink';
import { PyoliteKernelWorker } from './worker';
const worker = new PyoliteKernelWorker();
expose(worker);
```

- `worker.ts`
```ts
import { DriveFS } from '@jupyterlite/contents`;  // seems pretty legit?

export class PyoliteKernelWorker implements IPyoliteKernelWorker {
  // no constructor, if possible
  async initialize(options: PyoliteKernelWorker.IOptions): Promise<void> {
    this._options = options;
    await this.initRuntime(options);
    await this.initFS(options); // this is new!
    await this.initPackageManager(options);
    await this.initKernel(options);
    await this.initGlobals(options);
  }
}
```

If you like, I can try and take a crack at a sub-PR

martinRenou commented:
> though the pain of not being able to use `import` properly is starting to catch up, as I'd think we don't want the drivefs to have to live forever inside the kernel.

I totally agree, I was thinking the same thing. We've compiled xeus-python recently and made a JupyterLite kernel for it https://github.com/jupyterlite/xeus-python-kernel, I'd really want it to use the DriveFS too and there is no reason for the implementation to be different.

I'm actually tempted to remove the comlink usage from this PR and move it to a separate branch, as it's kind-of parallel to the DriveFS work. I am not sure we can use comlink for the DriveFS unless comlink can be used with a service worker for making synchronous tasks.

bollwyvl commented:
> parallel to the DriveFS work

but it _is_ using DriveFS from inside comlink worker at this time? I guess I haven't traced through which layers are doing what.

bollwyvl commented:
Created https://github.com/martinRenou/jupyterlite/pull/1: it breaks a lot, but might have some salveageable ideas.

I'll be (all but) offline over the weekend, but happy to keep tinkering with this: i'd really rather see us get something many folks can build on and use... but am semi-concerned about some of the browser limitations.

martinRenou commented:
Thanks! I'll also be offline over the weekend and on Monday, will look at it as soon as I'm back.

martinRenou commented:
From @bollwyvl in https://github.com/martinRenou/jupyterlite/pull/1:

> - based on https://github.com/jupyterlite/jupyterlite/pull/655
> - inching towards https://github.com/jupyterlite/jupyterlite/issues/463
> - splits up more of the comlink, FS, etc. stuff
> - removes use of buildScripts to create worker
> - removes lots of declares
> - adds (some) more typing
> - very WIP, many things broken
>   - os.listdir()
>   - drive in firefox
>   - doesn't even load in firefox private browsing (navigator.serviceworker undefined)

I'll try to find why `os.listdir()` does not work anymore.

I've been testing it on Chromium only for now, will try later on Firefox as well. Concerning the Firefox private browsing issue, this seems to be a known issue https://bugzilla.mozilla.org/show_bug.cgi?id=1320796, we probably want to give some meaningful feedback to the user in that case.

martinRenou commented:
My latest commit doesn't work well for now. It requires this fix from pyodide https://github.com/pyodide/pyodide/pull/2582 to be included.

bollwyvl commented:
> we probably want to give some meaningful feedback to the user in that case.

i feel like, beyond feedback, it will _have_ to be some kind of graceful degredation, not a hard fail telling you to use your browser a different way. Likely, this is wholly separate kernels, with the service-worker-powered one a subclass of the more limited ephemeral one.

At some point in the not-too-distant future, emscripten's `WasmFS` will give us the right tools for the job, even without a service worker.

martinRenou commented:
Making some progress, I have support for:
- creating directories
- listing files/directories
- renaming files
- removing directories

![filesystem](https://user-images.githubusercontent.com/21197331/172407045-76342079-d7f6-45c2-ba27-c32cb28dd57a.gif)

There is a permission issue when trying to access files though, I need to investigate it.

Also there is an issue with the service worker, it seems to only work for the tab that installed the service worker, after that it does not intercept requests. I suspect it needs to claim new tabs when they are open.

martinRenou commented:
We can now read/write files! I took the https://github.com/gzuidhof/starboard-notebook repo as an inspiration for the implementation.

There are a couple of things I would like to fix/improve/check still, after that I think the PR is ready:
- [ ] Reading a file content adds quotes around the string value while it shouldn't
- [ ] The service worker works the first time it is installed, but refreshing the page breaks the communication (maybe the service worker does not claim new tabs, though it should)
- [ ] Check reading/writing binary data
- [ ] Move the "server" part (the one responsible for using the content manager) to its own plugin
- [ ] Remove the IContent exposition to the IKernel I added in the first commit, we don't need it
- [ ] Make use of the kernel "location" passed to the kernel and have the kernel do a `cd` to that location after mounting the drive

jtpio commented:
Nice progress!

>  Reading a file content adds quotes around the string value while it shouldn't

Random guess: maybe related to a `JSON.stringify` somewhere?

martinRenou commented:
Probably! I'll have a try, thanks for the hint

martinRenou commented:
Thanks, it was indeed the `JSON.stringify` that was adding the extra quote

martinRenou commented:
@jtpio @bollwyvl I rebased and marked the PR as ready for review. If you agree we can tackle the two remaining points I added above in separate PRs (and open issues to track that).

Here is a picture of loading an image I dragged/dropped in jupyterlite and loaded from a Notebook that is in a subdirectory:
![file_read](https://user-images.githubusercontent.com/21197331/172884036-5804ddf4-d3f1-43c8-b970-3eddb4b8b25c.png)

jtpio commented on packages/server-extension/src/index.ts:
```diff
@@ -187,6 +193,177 @@ const contentsRoutesPlugin: JupyterLiteServerPlugin<void> = {
+const serviceWorkerPlugin: JupyterLiteServerPlugin<void> = {
+  id: '@jupyterlite/server-extension:service-worker',
+  autoStart: true,
+  requires: [],
```
Looks like this could be removed if it's empty:

```suggestion
```

jtpio replied:
Also having the service worker logic in its own plugin sounds like a good idea :+1: 

Since it doesn't depend on other services and doesn't provide a token, it could easily be omitted or disabled (which could be useful to https://github.com/executablebooks/thebe-core/pull/3).

jtpio commented on packages/contents/src/drivefs.ts:
```diff
@@ -0,0 +1,517 @@
+    this.fs = fs;
+  }
+
+  public open(stream: IEmscriptenStream): void {
```
Maybe we could drop the `public` here (and below) for consistency with the rest of the code base:

```suggestion
  open(stream: IEmscriptenStream): void {
```

jtpio commented on packages/server-extension/src/index.ts:
```diff
@@ -187,6 +193,177 @@ const contentsRoutesPlugin: JupyterLiteServerPlugin<void> = {
+  autoStart: true,
+  requires: [],
+  activate: (app: JupyterLiteServer) => {
+    navigator.serviceWorker
```
Probably we could handle `navigator.serviceWorker` not being available, which seems to be the case in a Firefox private window.

So the plugin does not fail at startup.

bollwyvl commented on packages/contents/src/drivefs.ts:
```diff
@@ -0,0 +1,517 @@
+export const DIR_MODE = 16895; // 040777
```
a lot of this looks copied: probably needs some headers with URLs and license info from the sources.

bollwyvl commented on packages/contents/src/drivefs.ts:
```diff
@@ -0,0 +1,517 @@
+}
+
+/**
+ * Wrap serviceworker requests for an Emscripten-compatible syncronous API.
```
can go either way: i almost feel like using the code-style `ServiceWorker` is more appropriate.

bollwyvl commented on packages/pyolite-kernel/src/declarations.d.ts:
```diff
@@ -7,10 +7,3 @@ declare module '!!file-loader*' {
   const res: string;
   return res;
 }
-
```
these deletions feel great... i wonder if this whole file can be deleted now?

bollwyvl replied:
we do still need it, for now, for the file-loader stuff... but `raw` can go... as can `buildScripts`... having a look...

bollwyvl commented on packages/pyolite-kernel/src/worker.ts:
```diff
@@ -1,420 +1,575 @@
+    await this.initPackageManager(options);
+    await this.initKernel(options);
+    await this.initGlobals(options);
+    await this.initFilesystem(options);
```
Hm... I know i moved this down because i was breaking other stuff, but I wonder if we can move `initFilesystem` up to somwhere before `initKernel`: this would enable a _not_ not-invented-here way to do startup code for kernels, e.g. ipython profiles.

bollwyvl commented on packages/pyolite-kernel/src/worker.ts:
```diff
@@ -1,420 +1,575 @@
+    FS.mount(driveFS, {}, '/drive');
+    FS.chdir('/drive');
+    this._driveFS = driveFS;
+    console.log('changing dir to ', options.location);
```
would remove

bollwyvl commented:
Hm... been experimenting locally and on RTD on the other PR... on Linux with `Chromium 102.0.5005.61` it doesn't really seem like any of the drivefs stuff works in private or standard browsing... and firefox has some other issues.

martinRenou commented:
Thank you both for the reviews! I'll resolve the comments.

> Hm... been experimenting locally and on RTD on the other PR... on Linux with Chromium 102.0.5005.61 it doesn't really seem like any of the drivefs stuff works in private or standard browsing... and firefox has some other issues.

Indeed it fails to `chdir` to the kernel location at startup. I'm looking into it right now.

martinRenou commented on packages/server-extension/src/index.ts:
```diff
@@ -187,6 +193,177 @@ const contentsRoutesPlugin: JupyterLiteServerPlugin<void> = {
+      );
+
+    // Setup communication with service worker for the virtual fs
+    const broadcast = new BroadcastChannel('/api/drive.v1');
```
I agree, but it might be difficult for the `/services.js` to import such a string from another place?

martinRenou commented:
It should now be more robust when it comes to hard refreshing the page :)

Please give it a try on RTD! Also I made a proper ServiceWorker plugin as you suggested, and the pyolite kernel only mounts the custom drive if the service worker registration succeeded.

martinRenou commented on packages/contents/src/drivefs.ts:
```diff
@@ -0,0 +1,517 @@
+ * Wrap serviceworker requests for an Emscripten-compatible syncronous API.
+ */
+export class ContentsAPI {
+  constructor(baseUrl: string, FS: any, ERRNO_CODES: any) {
```
I'll guess we'll want to contribute those types to `@types/emscripten` actually, it doesn't have the types we need yet https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/emscripten/index.d.ts

bollwyvl commented:
Getting much further! It is impressive

Observations from running RTD: 
- `open` and `read`ing text and binary looks good... 
  - reading a notebook always returns `[object Object]`
    - ![Screenshot from 2022-06-10 09-08-03](https://user-images.githubusercontent.com/45380/173083790-07346b5c-a06e-4f05-88e3-d4122f76f833.png)
- when opening a second browser tab with `?room=` in the same (browser, incognito), the python kernel disappears from the launcher, without any console messages, and can't be selected from the kernel change menu
  - ![Screenshot from 2022-06-10 09-12-21](https://user-images.githubusercontent.com/45380/173084622-63f8bc5f-7de4-4d05-833e-7934a13edd1c.png)
    - subsequent reloads of the browser also missing the kernel, no kernel message
    - after dev tools -> application -> storage -> clear site data, python kernel appears again
- if a file with spaces in the name is drag-and-dropped into the file browser, the parent directory (but not subdirectories) starts to appear empty to python
  - ![Screenshot from 2022-06-10 09-26-12](https://user-images.githubusercontent.com/45380/173087228-28db88ec-d2c0-42ed-b9d4-d00510aecbd4.png)
  -  it does fire some SW requests for _every_ file, however
  - removing the space from the filename makes things work again

martinRenou commented:
One nice side effect of the virtual file system implementation: completion on file names!
![file_completion](https://user-images.githubusercontent.com/21197331/173094118-1e976922-fef8-4c1e-86db-7fd8b9448e4a.gif)

martinRenou commented:
Thanks for the previous testing @bollwyvl 

I pushed some fixes for:
- issue when file names contain spaces
- issue when reading JSON files
- issue with the Python kernel not showing up

martinRenou commented:
In combination with #666 which fixes a CSV upload issue we can open CSV files with pandas:

![csv](https://user-images.githubusercontent.com/21197331/173301875-4d2685b2-9726-426d-afc4-175bb59b5b1d.png)

jtpio commented:
This is looking good, thanks Martin!!

Do you think we could also update the documentation here: https://jupyterlite.readthedocs.io/en/latest/howto/content/python.html
It would also be fine to do that as a follow-up.

jtpio commented on app/services.js:
```diff
@@ -0,0 +1,58 @@
+const broadcast = new BroadcastChannel('/api/drive.v1');
+
+self.addEventListener('install', (event) => {
+  console.log('Install JupyterLite service v1');
```
Should we keep these `console.log()` here? (some site deployers want quiet apps)

jtpio commented on packages/server-extension/src/index.ts:
```diff
@@ -247,7 +423,7 @@ const kernelSpecRoutesPlugin: JupyterLiteServerPlugin<void> = {
   autoStart: true,
   requires: [IKernelSpecs],
-  activate: (app: JupyterLiteServer, kernelspecs: IKernelSpecs) => {
+  activate: async (app: JupyterLiteServer, kernelspecs: IKernelSpecs) => {
```
The diff shows the addition of `async`, but it looks like this is not needed?

jtpio reviewed (approved):
Thanks @martinRenou and @bollwyvl!!

This is solving a big usability issue and working well on the RTD preview :+1: 

We can still iterate in follow-up PRs and issues.

jtpio commented:
cc @bollwyvl if you want to take another look at it.

Happy to make a new release with this when it lands.

martinRenou commented:
I'd like to push a last commit so that it respects custom `drive`s

psychemedia commented:
This PR looks really exciting, not least because it seems to work with `pandas` csv file handling for local files at least.

I note that sqlite doesn't play nicely though:

```
import sqlite3

conn = sqlite3.connect('test.db')

conn.execute('''CREATE TABLE TEST
         (ID INT PRIMARY KEY     NOT NULL,
         COL           TEXT    NOT NULL);''')

conn.close()

#---
"""
---------------------------------------------------------------------------
OperationalError                          Traceback (most recent call last)
Input In [5], in <cell line: 3>()
      1 conn = sqlite3.connect('test.db')
----> 3 conn.execute('''CREATE TABLE TEST
      4          (ID INT PRIMARY KEY     NOT NULL,
      5          COL           TEXT    NOT NULL);''')
      7 conn.close()

OperationalError: disk I/O error
""'
```

martinRenou commented:
Thanks for testing it out! I'll have a look at this issue.

bollwyvl commented:
> respects custom `drive`s

This seems reasonable... is there a good one to add to the demo site? of the ones i know are maybe too complicated/opinionated:
- `jupyterlab-github` would be great (though proprietary), but the read-only UI acts kinda broken (even though it's working)
- `jupyterlab-filesystem-access` is maybe _not_ a good baseline, because of the only-works-in-chrome thing... 
- `jupyterlab-s3-browser` might always need a token
- `jupyterlab-hdf5` has a rather complicated extension

All of this makes me wonder if getting a working `libarchive` backend working would make sense for more than just #653: akin to `hdf5`, it could use the default browser to fetch the archive and then spawn a new idrive from that...

martinRenou commented:
I pushed another commit for drive support:

jupyterlab-github:
![file_github](https://user-images.githubusercontent.com/21197331/173596400-c06ac6e2-6124-4671-9682-9d5a257aed59.png)

jupyterlab-filesystem-access:
![file_local](https://user-images.githubusercontent.com/21197331/173596461-16ba132e-fbdc-4c22-b6b4-8a6cd1139272.png)

martinRenou commented on packages/contents/src/drivefs.ts:
```diff
@@ -384,29 +390,53 @@ export class ContentsAPI {
+  normalizePath(path: string): string {
+    // Remove drive prefix
+    // The following check should always be true but for safety we check
+    if (path.startsWith('/drive')) {
```
Yeah definitely 👍🏽

bollwyvl commented on packages/contents/src/drivefs.ts:
```diff
@@ -384,29 +390,53 @@ export class ContentsAPI {
+    }
+
+    if (this._driveName) {
+      path = `${this._driveName}:${path}`;
```
perhaps hoisting the separator, and making it multiple characters, e.g. `DRIVE_SEP = '::'`

martinRenou commented:
> This seems reasonable... is there a good one to add to the demo site? of the ones i know are maybe too complicated/opinionated

I personally think the jupyterlab-filesystem-access one would be a good candidate, but indeed only works with Chrome. Which is no big deal, Firefox users will just not see it.

martinRenou commented on packages/contents/src/drivefs.ts:
```diff
@@ -0,0 +1,551 @@
+    } else if (whence === SEEK_END) {
+      if (this.fs.FS.isFile(stream.node.mode)) {
+        try {
+          position += stream.file!.data.length;
```
Sorry I don't understand what you are saying, can you elaborate on this?

bollwyvl replied:
yeah: this is the single warning that shows up in eslint the command line and in the review ui. 


![Screenshot from 2022-06-14 20-06-01](https://user-images.githubusercontent.com/45380/173714654-0564af14-5ac2-4c7f-9f40-54e8724cfe06.png)


i'm thinking we do one of:
- fail if there's no file (this function already throws a lot) 
- console.error as continue happily
- ignore the linter rule (would prefer not)

martinRenou replied:
Thanks :) I went for the first proposition

psychemedia commented:
Good support for filesystem access would be much appreciated in edu settings; `jupyerlite` offers the promise of install free, which is a big plus, and being able to support local file system access in a simple way also provides a route to file persistence by individuals.

*(Note to self: check local filesystem access is fine with USB mounted directories.)*

martinRenou commented:
Thanks for the review again @bollwyvl. I resolved most of your comments, I have not understood one of them though.

If you don't mind we can try to tackle the issue @psychemedia reported in a follow-up PR.

psychemedia commented:
If I download a python file into memory and save it as a `.txt` file, I can click to open it as expected from the JupyterLab file browser.

If I save it as a `.py` file and then click to open it from the file browser, I see base64 encoded data.

```python
import pyodide

# Load in the contents of `ouseful_jupyterlite_utils/utils` package from a remote URL:

URL = "https://raw.githubusercontent.com/innovationOUtside/ouseful_jupyterlite_utils/main/ouseful_jupyterlite_utils/utils.py"
txt = pyodide.open_url(URL).read()

with open("test.txt", "w") as f:
    f.write(txt)
    f.close()
# Click to open file displays text as expected

with open("test.py", "w") as f:
    f.write(txt)
    f.close()
# Click to open file displays base64 encoded data
```

jtpio commented:
Thanks all!

Let's move forward and get this in so we can iterate and improve things in follow-up issues.

jtpio commented:
Let's try to get https://github.com/jupyterlite/jupyterlite/pull/670 as well before making a new release.

martinRenou commented:
You made my day merging this, thanks!!
